### PR TITLE
🌱 controller/machine: use unstructured caching client

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

### tl;dr

I think we can cache all our get calls with unstructured in the Machine controller. This leads to huge performance improvements at scale (an average Machine reconcile goes from ~ 1 second to double digits milliseconds).

Most of this PR is just updating the tests.

### Why do I think it's safe to cache the unstructured gets?

While it could happen that a Machine reconcile is seeing stale BootstrapConfigs / InfraMachines, based on experiments and upstream documentation for every update on BootstrapConfig / InfraMachine we will get another Machine reconcile. During that reconcile the BootstrapConfig / InfraMachine will have been already updated in the cache.

This is the case because Kubernetes informers always first update the cache before they notify event handlers. (Event handlers eventually enqueue a reconcile request for our Machines)

client-go-controller-interaction
(Source: https://github.com/kubernetes/sample-controller/blob/master/docs/controller-client-go.md)

In this diagram it can be seen that once an event is received in 1), the informer always first updates the cache in 5) before triggering the event handler in 6). A controller-runtime controller reconciles roughly after 8).

Please note that there is no way to guarantee that a Machine reconcile will always use a 100% up-to-date BootstrapConfig / InfraMachine as it's impossible to guarantee that the reconciler gets the objects which might have been written during the Machine reconcile - even with a live client.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related #8814

### CODE REVIEW

**Comment by sbueringer** [comment]
I'll take a look at the e2e tests.

**Comment by richardcase** [comment]
This is great @sbueringer . In our scale testing, we are seeing the read performance of `Machine` as one of the biggest bottlenecks. Above about 500 clusters, we see that the read calls start exceeding the SLO of 1s. So if this change brings it down to less than 100ms that is fantastic :tada:

**Comment by fabriziopandini** [comment]
/lgtm

I think it is ok to use cached read for external objects in the machine controller, given that as explained above we are watching for those objects and the machine controller will reconcile again as soon as boostrapConfig or InfraMachine will change

Looking at this from another perspective, I would say the current implementation is more due to the limitations of the cache (or of the cache options) at the time we first implemented this code; over time not only controller runtime has been improved a lot (kudos to the team), but we have a better and deep understanding of how all this work and we can now make changes like this in a very surgical way.

Note: eventually, in a follow up we can optimize the memory footprint of the cache for boostrapConfig and InfraMachine by keeping in memory only the subset of object fields that are defined by the contract and dropping everything else.

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
